Experimental development-only single-pass webpack workflow

kevin940726 opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
Part of https://github.com/WordPress/gutenberg/discussions/36142.

Currently, our development workflow works roughly like this:
1. `npm run dev`: which runs (2), (3) and (4).
2. `npm run build:packages`: build all packages with Babel for js/ts and SASS for css into _package-level_ build directories (`build`, `build-module`, `build-style`) once.
3. `npm run dev:packages`: start watchers to watch for all the source files and rebuild (2) when they change.
4. `wp-scripts start`: run webpack watch mode and watch for the **build** files built by (2) and (3) then bundle them to _root-level_ `build/` directory.
5. Browsers import files in (4).

The problem is that whenever we change something in our source code, we have to wait for Babel/SASS to rebuild them into _package-level_ builds (3), and then wait until webpack pick them up and bundle them into the final build. It's slow for both initial build and incremental build, since that a lot of expensive IO are wasted on building intermediate results. While we do need the intermediate _package-level_ builds for production and deployment, we don't need them during development.

This PR tries to simplify that by leveraging webpack to do all the heavy lifting during development. It'll handle all the JS files along with all the CSS files, directly from their sources. We won't be outputting intermediate builds, no custom build scripts and extra watchers. The build performance should be a lot faster for initial build and more responsive for incremental builds too.

This PR introduces a new command `npm run dev@exp` to run the experimental `dev` build. It runs on a different path than the original `dev` command behind a feature flag, so that we can safely test it without risking breaking everyone's build.

I ran a couple of builds locally to compare between the two, and below is what I saw on my macbook pro 2020 (2.3GHz Intel Core i7, 16GB RAM, macOS 12.0.1). (Since that both commands will start watchers and never finish, we have to do some some tweaks to be able to run the `time` audit. First, go to `packages/scripts/scripts/start.js` and remove `'--watch'`. Then, for `npm run dev`, you'll have to remove `npm run dev:packages` in the scripts in `package.json`. Finally, for `npm run dev@exp`, you'll have to comment out `pollTimeout: ...` in `tools/webpack/experimental.config.js`) (You might want to run each the second time to take cached/warmed results into account.)

**For initial build:**

```
# time npm run dev
225.23s user 23.39s system 339% cpu 1:13.22 total

# time npm run dev@exp
48.61s user 5.65s system 185% cpu 29.255 total
```

The total amount of time is reduced to about **40%** and the CPU usage is only **55%** too. 

**For incremental builds:**

I don't have a consistent setup for auditing incremental build performance though, so below is just a rough number in my local testing:

`dev` | `dev@exp`
--- | ---
~3s | ~1.5s

It's still only around **50%** of the incremental build time though.

Note that this change isn't only for performance, but it also makes implementing other enhancements in https://github.com/WordPress/gutenberg/discussions/36142 easier in the future, performance boost is just a bonus.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
1. `npm run dev` should work the same
2. `npm run dev@exp` should also work as expected
3. Do a fresh clone of the repo and run `npm install` and `npm run dev@exp` should also work

## Screenshots <!-- if applicable -->

![image](https://user-images.githubusercontent.com/7753001/144191850-af8f4e14-0f4b-4338-ab66-a00a1d8f5a34.png)


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New feature

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

kevin940726 committed:
Add experimental dev-only webpack single pass build

adamziel commented:
I wonder if this could replace webpack entirely in the longer run: https://swc.rs/

kevin940726 commented:
> I wonder if this could replace webpack entirely in the longer run: https://swc.rs/

That could replace Babel, it's mentioned in the original discussion in #36142.

noisysocks commented:
> Size Change: -240 kB (-18%) 👏

Wasn't expecting this. Did it mess up the Size Change script?

kevin940726 commented:
> Wasn't expecting this. Did it mess up the Size Change script?

Fixed!

gziolo commented:
Thank you for starting this PR. It looks like the direction in the JavaScript community is to further decouple the development process from production builds. Vite is the best example for that. 

> **For initial build:**
> 
> ```
> # time npm run dev
> 152.03s user 15.98s system 435% cpu 38.564 total
> 
> # time npm run dev@exp
> 27.25s user 3.18s system 148% cpu 20.496 total
> ```
> 
> The total amount of time is reduced to nearly **50%** and the CPU usage is only **1/3** too. 

`npm run dev` includes also additional steps to build and verify types with TypeScript which might have significant impact on the comparison:

https://github.com/WordPress/gutenberg/blob/88883a3ea23f5b3d8aff563f6a8447e4133b9e55/package.json#L228

It would be great to measure it. Also, are we sure we don’t need want to bother with types during development? I’m wondering what the experience would be in case someone works on changes to the packages that require modifications to TypeScript definitions.

kevin940726 commented:
@gziolo Thank you for mentioning that! I added `fork-ts-checker-webpack-plugin` to help do type checking during development. This is also a new feature since that I don't think we're checking TS types for incremental builds currently.

kevin940726 commented on packages/list-reusable-blocks/package.json:
```diff
@@ -22,6 +22,7 @@
 	"engines": {
 		"node": ">=12"
 	},
+	"gutenberg:source": "src/index.js",
```
Ideally, we should add this to all packages so that we don't have to use `react-native` as the main field.

Or we can write our own custom resolver, not sure if it's worth it though.

kevin940726 commented on tools/webpack/experimental.config.js:
```diff
@@ -0,0 +1,423 @@
+	mode,
+	devtool:
+		devtool ||
+		( mode === 'production' ? false : 'eval-cheap-module-source-map' ),
```
For some reasons, source maps for CSS only works if we use `source-map` here. I guess it could be an opt-in thing for now if folks really need source maps for them.

kevin940726 committed:
Only warm up thread-loader in development

kevin940726 committed:
Add wp-polyfill for packages builds

kevin940726 commented on tools/webpack/experimental.config.js:
```diff
@@ -0,0 +1,425 @@
+				build: true,
+			},
+		} ),
+		new IgnoreEmitPlugin( /\.css-entry\.js$/ ),
```
We're ignoring the unused CSS entry js for now since they serve no purposes. However, in the future, we could use them for hot reloading CSS during development.

kevin940726 committed:
Only set infinity poll timeout in development

noisysocks commented:
When testing locally on an already very fast computer I see a ~2x speed up (15s → 7s) for initial build and a ~4x(!!) speed up (2s → 0.5s) for incremental build. Definitely will improve some lives!

I like that it's a seperate `npm run` command so that we can test this in real world conditions first to see if there are issues. I'd say if GitHub Actions is happy then let's give it a go.

kevin940726 commented:
@gziolo Would you mind giving it another round of review? Thanks!

gziolo commented:
In https://github.com/WordPress/gutenberg/pull/36736, @youknowriad explores speeding up the build by splitting the tasks into individual packages and using SWC instead of Babel. It doesn't necessarily mean this is an alternative for this PR, but we should explore a wider set of options before we jump to the conclusion of how we should improve the development workflows.

youknowriad commented:
The approach in this PR has a risk of hidden bugs because of the different build pipelines between packages and plugin. I think if we can speed up the packages build and optimize the bundling step, it'd be an ideal approach.

The other thing I came up with is that we might not be able to assume that all packages are built in the exact same way in the future, and having the flexibility for each package to define its own way to build is a good thing (in addition to sharing common building using wp-scripts or similar)

kevin940726 commented:
> The approach in this PR has a risk of hidden bugs because of the different build pipelines between packages and plugin. I think if we can speed up the packages build and optimize the bundling step, it'd be an ideal approach.

Why? We're already building different builds between packages and plugin, just that the plugin build depends on the packages build, the final output is still different. Furthermore, this PR only targets development build, so the production build is still intact.

However, IMO plugin and packages build should still have different pipelines, as long as they are using the same source files. Which build pipeline they use, whether it is webpack, babel, or SWC, is implementation detail, and it shouldn't matter for the final results. It's even better as we can apply different optimizations for different pipelines. For instance, JSX transpilation can continue to use `createElement` for package builds, and use the new automatic runtime for plugin build. Another example is that webpack can parse dynamic imports differently and enable code splitting and lazy loading for those assets.

> The other thing I came up with is that we might not be able to assume that all packages are built in the exact same way in the future, and having the flexibility for each package to define its own way to build is a good thing (in addition to sharing common building using wp-scripts or similar)

I agree, the `gutenberg:source` field in `package.json` I added in this PR is an example of that. I don't think we need to specify all the details though, for most packages we only just need the source entry files. We can add some other fields for other special packages if necessary.

youknowriad commented:
> Why? We're already building different builds between packages and plugin, just that the plugin build depends on the packages build, the final output is still different. Furthermore, this PR only targets development build, so the production build is still intact.

The main issue for me is that Core consumes the production build but I've been thinking more and I think it's not that important especially since our CI also uses the production build. So I wouldn't be against a separate pipeline. 

That said, I wonder if we might be able to get good optimizations for both production and dev tooling potentially if we optimize the way we build the deps (see tools like rush, turbo, nx) while avoiding the new build tool to maintain (dev specific one).

Anyway, I think our north star in terms of build tooling should be to avoid custom tooling and custom things in general to retain as much flexibility as we can and avoid locking. (custom loaders, babel plugins...) There's a lot of innovation in this space at the moment and we shouldn't prevent ourselves from trying different approaches.

kevin940726 commented:
> That said, I wonder if we might be able to get good optimizations for both production and dev tooling potentially if we optimize the way we build the deps (see tools like rush, turbo, nx) while avoiding the new build tool to maintain (dev specific one).

I agree! I actually tried hard to make this PR work for production too, but I haven't tested it to be sure. So in the future, hopefully we can optimize production builds with the help of this PR too.
 
> Anyway, I think our north star in terms of build tooling should be to avoid custom tooling and custom things in general to retain as much flexibility as we can and avoid locking. (custom loaders, babel plugins...) There's a lot of innovation in this space at the moment and we shouldn't prevent ourselves from trying different approaches.

That's also my goal with this PR! Most modern build tools nowadays expect a single-pass pipeline to optimize the performance. If this PR works, then I believe we just prove that we don't need special handling or multiple pipelines. By returning to the basic, common way of configuring our webpack configs, I believe we can also make the migration path smoother if we ever want to move to other build tools. webpack is still the most popular build tool for web apps right now, so that we can expect next-gen build tools to provide smoother migration path from it.

What do you say if we merge this PR and try it out? The whole thing is still behind a flag though so we can experiment more stuffs on it. I have other optimizations planned ahead (#36142), but we need this to be merged to be able to work on those stuffs 😅 .

youknowriad commented:
>  So in the future, hopefully we can optimize production builds with the help of this PR too.

Actually, I think we should always stick with the two steps build for production, because that's how Core consumes the packages (it consumes the built packages) and we shouldn't derail from that otherwise we risk having bugs that only show up in Core but not Gutenberg.

> What do you say if we merge this PR and try it out? 

not against trying this but not sure we need the flag, everyone working in the project should experience it, we can always revert or whatever.

gziolo commented:
As of today, the development build process `npm run dev` has a few steps like @youknowriad mentioned. I did a quick meassurement of the impact each step has:

1. Remove all build artifacts (excluding types) and run a local build process for a few packages:
    https://github.com/WordPress/gutenberg/blob/1d5ba84192e5442cc9efa984763b07b0e5041173/package.json#L228
    > npm run prebuild:packages  2.25s user 1.50s system 147% cpu 2.539 total
2. Build TypeScript types for packages (it is incremental, so it will take way more time on fresh install): 
    https://github.com/WordPress/gutenberg/blob/1d5ba84192e5442cc9efa984763b07b0e5041173/package.json#L230
    > npm run build:package-types  2.30s user 0.14s system 142% cpu 1.712 tota
3. Build JS and CSS for all packages (`node ./bin/packages/build.js` part):
    https://github.com/WordPress/gutenberg/blob/1d5ba84192e5442cc9efa984763b07b0e5041173/package.json#L229
    > node ./bin/packages/build.js  56.76s user 5.72s system 880% cpu 7.100 total
4. Run the development build with webpack:
    https://github.com/WordPress/gutenberg/blob/1d5ba84192e5442cc9efa984763b07b0e5041173/package.json#L237
    > npx wp-scripts start  0.07s user 0.01s system 1% cpu 7.605 total

This PR proposes that we remove steps (1) and (2) plus we combine steps (3) and (4). It's obvious that the part where Babel and PostCSS are used to generate packages (3) is the most resource-consuming. I would be curious to see the breakdown between Babel and CSS processing (SASS, PostCSS) as well. For the time being, we can mostly plan to swap Babel with SWC, but I'm not aware of any replacement for faster CSS processing.

It's also worth noting that the current webpack config sets non-standard settings for watch mode:

https://github.com/WordPress/gutenberg/blob/1d5ba84192e5442cc9efa984763b07b0e5041173/tools/webpack/shared.js#L54-L60

The default is 200ms for `aggregateTimeout` so by design, there is a 300ms difference between the existing command and the one proposed in this PR.

gziolo commented:
Actually, it isn't that hard to compare the time spent by `node ./bin/packages/build.js` on processing CSS vs JS files.

1. JavaScript and TypeScript files:
    > node ./bin/packages/build.js  44.56s user 4.78s system 841% cpu 5.864 total
2. SCSS files:
    > node ./bin/packages/build.js  12.11s user 1.16s system 798% cpu 1.660 total

It's also worth mentioning, that the way packages are published to npm is shaped to support also Common JS, so we have two output folders: `build` (CJS) and `build-module` (ESM).

If we skip the CJS that isn't used by webpack:

```diff
diff --git a/bin/packages/build-worker.js b/bin/packages/build-worker.js
index a984e96b6c..d05916dfd0 100644
--- a/bin/packages/build-worker.js
+++ b/bin/packages/build-worker.js
@@ -28,7 +28,7 @@ const PACKAGES_DIR = path
  * @type {Object}
  */
 const JS_ENVIRONMENTS = {
-       main: 'build',
+       // main: 'build',
        module: 'build-module',
 };
```

then we get another performance improvement when building both CSS and JS:

> node ./bin/packages/build.js  35.70s user 3.60s system 872% cpu 4.504 total

I think we can improve the measured performance of the existing development workflow by 30-40% only by aligning the tasks with what this PR proposes. The fact that most of the CPU time is spent on building packages, we should also see a big reduction when doing only one pass for ES Modules.

kevin940726 commented:
> Actually, I think we should always stick with the two steps build for production, because that's how Core consumes the packages (it consumes the built packages) and we shouldn't derail from that otherwise we risk having bugs that only show up in Core but not Gutenberg.

Why can't we change how it works in Core too? I mean just let Core consumes the same source code as in gutenberg.

> ...then we get another performance improvement when building both CSS and JS...

True, that's what this PR does, but this PR also refactored and cleaned up the build process for future enhancement. It's also worth noting that the CPU usage is way lower if we only use webpack to build our package, and we don't need the _non-standard settings for watch mode_ either.

It's even stated in the [official doc](https://webpack.js.org/guides/build-performance/#incremental-builds) of webpack too:

> Use webpack's watch mode. Don't use other tools to watch your files and invoke webpack. The built-in watch mode will keep track of timestamps and passes this information to the compilation for cache invalidation.

kevin940726 commented:
> not against trying this but not sure we need the flag, everyone working in the project should experience it, we can always revert or whatever.

Yeah, but I have other PRs planned after this is merged, and I don't want to risk blocking everyone and reverting this PR when I'm just experimenting with all these. I think it'll be safer and smoother for me to just land this behind a flag for now.

kevin940726 committed:
Fix package-lock

kevin940726 commented:
@youknowriad, @gziolo, would any of you like to give it another review 🙂 ?

gziolo commented:
> would any of you like to give it another review 🙂 ?

I would like to hear feedback from a wider group of people. We need not only to test this new development workflow as part of the review process, but we need to have volunteers who will ensure that it works smoothly daily when developing features in Gutenberg.

noisysocks commented:
> > would any of you like to give it another review 🙂 ?
> 
> I would like to hear feedback from a wider group of people. We need not only to test this new development workflow as part of the review process, but we need to have volunteers who will ensure that it works smoothly daily when developing features in Gutenberg.

Perhaps let's merge this and @kevin940726 can write a Make/Core post advertising the fact that we're looking for people to help test `npm run dev@exp` in their day-to-day workflows.

kevin940726 commented:
I haven't done anything to pick this up again but just want to share a diagram I created the other day that could potentially make this approach easier to understand.

Currently, our development workflow for JavaScript files roughly looks like this:
![Before](https://user-images.githubusercontent.com/7753001/230892149-d07136e0-5c0b-476a-be93-7f5134905813.png)

Each source file will go through all these steps to become available for the server to serve it. Each square has a cost, whether it is file system IO cost, or transpilation cost. Red arrows indicates the release process. You can see that we mostly need all those steps for the release workflow, but not required for the development workflow.

What this PR is trying to achieve is:
![After](https://user-images.githubusercontent.com/7753001/230892614-9cbcb139-f550-4b8c-a906-de5a1d453057.png)

We can get rid of most of the squares here and directly transpile and serve the built files using webpack. This is the end goal though, and I understand it might be too extreme to completely shift to that in one PR. Maybe we can do this in steps, but we should aim at the above diagram as the end result.

youknowriad commented:
Great work on the diagram @kevin940726 I think we should definitely save a version of it in the repo and add it to the "architecture" docs. 

- Maybe clarify that there are two targets: the "packages/*/build/*" that produces the npm packages
- and the the `build/*` that produces the WP scripts and styles.

kevin940726 commented:
@youknowriad Yeah, that's the dotted "group"s named "packages" and "plugin" are for. :) Open to ideas of how to make that clearer too! FWIW, the diagram is intentionally vague to only cover the things related to this PR. We can definitely refine it so that it can be served as a visual guide in the docs though!
